### PR TITLE
[FEATURE] bootstrap estimator  for NumericMetricRangeMultiBatchParameterBuilder

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Custom `bootstrap` estimator (utilized in `NumericMetricRangeMultiBatchParameterBuilder`)
* Unit test covering several distributions (normal, bimodal, exponential)

### Remarks
* This custom `bootstrap` does not implement any optimization (such as bias correction, etc.) available in `scipy.stats.bootstrap`; however, the latter requires `scipy` version `1.7.0` which is not available in all distributions.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [ ] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [ ] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [ ] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 1]
```diff
@@ -64,14 +63,14 @@ def __init__(
         sampling_method: Optional[str] = "bootstrap",
         enforce_numeric_metric: Optional[Union[str, bool]] = True,
         replace_nan_with_zero: Optional[Union[str, bool]] = True,
-        confidence_level: Optional[Union[float, str]] = 9.5e-1,
-        num_bootstrap_samples: Optional[Union[int, str]] = None,
-        round_decimals: Optional[Union[int, str]] = None,
+        confidence_level: Optional[Union[str, float]] = 9.5e-1,
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: rename to "false_positive_rate"; should be the desired false positive rate. We're no longer exposing confidence interval in this implementation (though we could estimates of parameter quality in meta (in a future PR))

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 2]
```diff
@@ -85,7 +84,7 @@ def __init__(
             enforce_numeric_metric: used in MetricConfiguration to insure that metric computations return numeric values
             replace_nan_with_zero: if False, then if the computed metric gives NaN, then exception is raised; otherwise,
             if True (default), then if the computed metric gives NaN, then it is converted to the 0.0 (float) value.
-            confidence_level: user-configured fraction between 0 and 1
+            confidence_level: user-configured fraction between 0 and 1 expressing desired confidence level in estimation
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]:
```suggestion
            false_positive_rate: user-configured fraction between 0 and 1 expressing desired false positive rate for identifying unexpected values, as judged by the upper- and lower- quantiles of the observed metric data.
```

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 3]
```diff
@@ -235,19 +199,53 @@ def _build_parameters(
...
+            in NumericMetricRangeMultiBatchParameterBuilder.RECOGNIZED_SAMPLING_METHOD_NAMES
+        ):
+            raise ge_exceptions.ProfilerExecutionError(
+                message=f"""The directive "sampling_method" for {self.__class__.__name__} can be only one of 
+{NumericMetricRangeMultiBatchParameterBuilder.RECOGNIZED_SAMPLING_METHOD_NAMES} ("{sampling_method}" was detected).
+"""
+            )
+
+        # Obtain confidence_level from rule state (i.e., variables and parameters); from instance variable otherwise.
+        confidence_level: Union[
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: rename from above

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 4]
```diff
@@ -321,61 +325,52 @@ def _get_bootstrap_confidence_interval(
...
-        confidence_interval: ConfidenceInterval = bootstrap_result.confidence_interval
-
-        confidence_interval_low: np.float64 = confidence_interval.low
-        confidence_interval_high: np.float64 = confidence_interval.high
-
-        std: Union[np.ndarray, np.float64] = bootstrap_result.standard_error
-
-        stds_multiplier: np.float64 = NP_SQRT_2 * special.erfinv(confidence_level)
-        margin_of_error: np.float64 = stds_multiplier * std
+        return self._compute_bootstrap_estimation_and_return_confidence_interval(
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Please remove the nested call.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 5]
```diff
@@ -321,61 +325,52 @@ def _get_bootstrap_confidence_interval(
...
-        confidence_interval_low -= margin_of_error
-        confidence_interval_high += margin_of_error
+    @staticmethod
+    def _compute_bootstrap_estimation_and_return_confidence_interval(
+        metric_values: np.ndarray,
+        confidence_level: np.float64,
+        n_resamples: int,
+    ):
+        uncertainty_level: float = 1.0 - confidence_level
+        quantile_probability_lower: float = 5.0e-1 * uncertainty_level
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]:
```suggestion
        quantile_probability_lower: float = uncertainty_level / 2
```

Please improve code readability for everyone.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 6]
```diff
@@ -321,61 +325,52 @@ def _get_bootstrap_confidence_interval(
...
+        quantile_probability_upper: float = 1.0 - quantile_probability_lower
 
-        return ConfidenceInterval(
-            low=confidence_interval_low, high=confidence_interval_high
+        bootstraps: np.ndarray = np.random.choice(
+            metric_values, size=(n_resamples, metric_values.size)
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: we may want to consider `min(metric_values.size, 100)` for the second dimension (FOR A FUTURE PR); this is fine.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 7]
```diff
@@ -321,61 +325,52 @@ def _get_bootstrap_confidence_interval(
...
+        )
+        high: np.float64 = np.mean(
+            np.quantile(
+                bootstraps,
+                axis=1,
+                q=quantile_probability_upper,
+                interpolation="linear",  # can be omitted ("linear" is default)
+            )
         )
+        return ConfidenceInterval(low=low, high=high)
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: This is no longer a confidence interval. Please return simply (low, high) (and remove all other references to confidence_interval per notes above.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 8]
```diff
@@ -321,61 +325,52 @@ def _get_bootstrap_confidence_interval(
...
 
     def _get_oneshot_confidence_interval(
         self,
-        metric_values: Union[
-            np.ndarray,
-            List[Union[int, np.int32, np.int64, float, np.float32, np.float64]],
-        ],
+        metric_values: Union[np.ndarray, List[Number]],
```
Comment (jcampbell) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Please replace onehshot with a straight-up quantile calculation.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py [hunk 9]
```diff
@@ -22,3 +48,31 @@ def test_standard_deviation_band_around_mean_rule():
     confidence_level = 9.97300203937e-1
     stds_multiplier = NP_SQRT_2 * special.erfinv(confidence_level)
     assert np.isclose(stds_multiplier, 3.0)
+
+
+def test_custom_bootstrap_efficacy():
```
Comment (jcampbell) [inline tests/rule_based_profiler/parameter_builder/test_parameter_computations.py:32]: YAY!

**Review by jcampbell — CHANGES_REQUESTED** [review]
Changes requested -- looking close!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: @jcampbell I need to run it from a test module; I can move it to the util module in order to do that.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Moved to util.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 8 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Done!

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 9 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py:32]: 🙇

**Review by jcampbell — APPROVED** [review]

**Pull request merged**